infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStorageProvisioner.java

```diff
@@ -0,0 +1,370 @@
+    String containerName = Names.generateName(ASYNC_STORAGE);
+
+    Volume storageVolume =
+        new VolumeBuilder()
+            .withName(STORAGE_VOLUME)
+            .withPersistentVolumeClaim(
+                new PersistentVolumeClaimVolumeSourceBuilder()
+                    .withClaimName(configuredPVCName)
+                    .withReadOnly(false)
+                    .build())
+            .build();
+
+    Volume sshKeyVolume =
+        new VolumeBuilder()
+            .withName(CONFIG_MAP_VOLUME_NAME)
+            .withConfigMap(
+                new ConfigMapVolumeSourceBuilder()
+                    .withName(configMap)
+                    .withDefaultMode(0600)
+                    .build())
+            .build();
+
+    VolumeMount storageVolumeMount =
+        new VolumeMountBuilder()
+            .withMountPath(ASYNC_STORAGE_DATA_PATH)
+            .withName(STORAGE_VOLUME)
+            .withReadOnly(false)
+            .build();
+
+    VolumeMount sshVolumeMount =
+        new VolumeMountBuilder()
+            .withMountPath(SSH_KEY_PATH)
+            .withSubPath(AUTHORIZED_KEYS)
+            .withName(CONFIG_MAP_VOLUME_NAME)
+            .withReadOnly(true)
+            .build();
+
+    Container container =
+        new ContainerBuilder()
+            .withName(containerName)
+            .withImage(storageImage)
+            .withNewResources()
+            .addToLimits("memory", new Quantity("512Mi"))
+            .addToRequests("memory", new Quantity("256Mi"))
+            .endResources()
+            .withPorts(
+                new ContainerPortBuilder()
+                    .withContainerPort(SERVICE_PORT)
+                    .withProtocol("TCP")
+                    .build())
+            .withVolumeMounts(storageVolumeMount, sshVolumeMount)
+            .build();
+
+    PodSpecBuilder podSpecBuilder = new PodSpecBuilder();
+    PodSpec podSpec =
+        podSpecBuilder.withContainers(container).withVolumes(storageVolume, sshKeyVolume).build();
+
+    return new PodBuilder()
+        .withApiVersion("v1")
+        .withKind("Pod")
+        .withNewMetadata()
+        .withName(ASYNC_STORAGE)
+        .withNamespace(namespace)
+        .withLabels(singletonMap("app", ASYNC_STORAGE))
+        .endMetadata()
+        .withSpec(podSpec)
+        .build();
+  }
+
+  /** Create service for serving rsync connection */
+  private Service createStorageService(String namespace) {
+    ObjectMeta meta = new ObjectMeta();
+    meta.setName(ASYNC_STORAGE);
+    meta.setNamespace(namespace);
+
+    IntOrString targetPort =
+        new IntOrStringBuilder().withIntVal(SERVICE_PORT).withStrVal(valueOf(SERVICE_PORT)).build();
+
+    ServicePort port =
+        new ServicePortBuilder()
+            .withName("rsync-port")
+            .withProtocol("TCP")
+            .withPort(SERVICE_PORT)
+            .withTargetPort(targetPort)
+            .build();
+    ServiceSpec spec = new ServiceSpec();
```

Comment:
[ServerServiceBuilder](https://github.com/eclipse/che/blob/master/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/ServerServiceBuilder.java#L61) can help you reduce the code

Reply:
https://github.com/eclipse/che/issues/17399